File: basic_calculator.cpp
#include "leetcode.h"

class Solution {
public:
  int calculate(string s) {
    stack<pair<int, int>> sp;
    long long int ans = 0;
    int sign = +1;
    for (int i = 0; i < s.size(); ++i) {
      char ch = s[i];
      if (isdigit(ch)) {
        long long int num = 0;
        while (i < s.size() && isdigit(s[i])) {
          num = (num * 10) + s[i] - '0';
          i++;
        }
        i--;
        ans += (num * sign);
        sign = +1;
      } else if (ch == '(') {
        sp.push(make_pair(ans, sign));
        ans = 0;
        sign = +1;
      } else if (ch == ')') {
        ans = sp.top().first + (sp.top().second * ans);
        sp.pop();
      } else if (ch == '-')
        sign = (-1 * sign);
    }
    return ans;
  }
};

int main() {
  Solution obj;
  cout << obj.calculate("1 + 1");
}

File: other/basic_calculator.rs
struct Solution;

impl Solution {
    pub fn calculate(s: String) -> i32 {
        let mut sp: Vec<i32> = Vec::new();
        let (mut ans, mut num, mut sign) = (0, 0, 1);
        sp.push(sign);
        for ch in s.chars() {
            match ch {
                ' ' => {}
                '+' | '-' => {
                    ans += num * sign;
                    sign = sp.last().unwrap() * if ch == '+' {1} else {-1};
                    num = 0;
                }
                '(' => {sp.push(sign);}
                ')' => {sp.pop();}
                _ => {num * 10 + (ch as i32 - '0' as i32)}
            };
        }
        ans += num * sign;
        ans
    }
}

fn main() {
    let s = String::from("1 + 1");
    print!("{}", Solution::calculate(s));
}
